pkg/serviceaccounts/controllers/docker_registry_service.go

```diff
@@ -0,0 +1,203 @@
+				return e.client.Services(options.RegistryNamespace).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Service{},
+		options.Resync,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    e.serviceAdded,
+			UpdateFunc: e.serviceUpdated,
+			DeleteFunc: e.serviceDeleted,
+		},
+	)
+	e.registryServiceName = options.RegistryServiceName
+	e.dockercfgController = options.DockercfgController
+
+	return e
+}
+
+// DockerRegistryServiceController manages ServiceToken secrets for Service objects
+type DockerRegistryServiceController struct {
+	stopChan chan struct{}
+
+	client client.Interface
+
+	registryServiceName string
+
+	dockercfgController *DockercfgController
+
+	serviceController *framework.Controller
+}
+
+// Runs controller loops and returns immediately
+func (e *DockerRegistryServiceController) Run() {
+	if e.stopChan == nil {
+		e.stopChan = make(chan struct{})
+		go e.serviceController.Run(e.stopChan)
+	}
+}
+
+// Stop gracefully shuts down this controller
+func (e *DockerRegistryServiceController) Stop() {
+	if e.stopChan != nil {
+		close(e.stopChan)
+		e.stopChan = nil
+	}
+}
+
+// serviceAdded reacts to the creation of a docker-registry service by updating all service account dockercfg secrets and
+// changing all interestedURLs
+func (e *DockerRegistryServiceController) serviceAdded(obj interface{}) {
+	service := obj.(*api.Service)
+	if service.Name != e.registryServiceName {
+		return
+	}
+
+	serviceLocation := service.Spec.PortalIP
+	if len(serviceLocation) == 0 {
+		serviceLocation = OpenshiftDockerURL
+	}
+
+	if err := e.handleLocationChange(serviceLocation); err != nil {
+		glog.Error(err)
+	}
+}
+
+// serviceUpdated reacts to the creation of a docker-registry service by updating all service account dockercfg secrets and
+// changing all interestedURLs, if needed
+func (e *DockerRegistryServiceController) serviceUpdated(oldObj interface{}, newObj interface{}) {
+	oldService := oldObj.(*api.Service)
+	newService := newObj.(*api.Service)
+	if newService.Name != e.registryServiceName {
+		return
+	}
+	if oldService.Spec.PortalIP == newService.Spec.PortalIP {
+		return
+	}
+
+	serviceLocation := newService.Spec.PortalIP
+	if len(serviceLocation) == 0 {
+		serviceLocation = OpenshiftDockerURL
+	}
+
+	if err := e.handleLocationChange(serviceLocation); err != nil {
+		glog.Error(err)
+	}
+}
+
+// serviceDeleted reacts to a Service deletion by deleting all corresponding ServiceToken Secrets
+func (e *DockerRegistryServiceController) serviceDeleted(obj interface{}) {
+	service := obj.(*api.Service)
+	if service.Name != e.registryServiceName {
+		return
+	}
+
+	if err := e.handleLocationChange(OpenshiftDockerURL); err != nil {
+		glog.Error(err)
+	}
+}
+
+// createDockercfgSecretIfNeeded makes sure at least one ServiceToken secret exists, and is included in the service's Secrets list
+func (e *DockerRegistryServiceController) handleLocationChange(serviceLocation string) error {
```

Comment:
> seems like this whole function belongs in dockercfgController, since it knows the nature of the annotations and contents of the secrets it creates

Disagree.  I'd rather have separate controllers for separate resources that coordinate based on data.  Having an understanding of the semantic meaning of the fields doesn't strike me as a significant leak.